### PR TITLE
refactor(lsp/rpc): move transport logic to separate module

### PR DESCRIPTION
**Pull request by lewis6991**
Simplifies `vim.lsp.rpc.start()` and `vim.lsp.rpc.connect()` and puts them both into very similar forms.

### CODE REVIEW

File: runtime/lua/vim/lsp/_transport.lua [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+local log = require('vim.lsp.log')
+
+local is_win = vim.fn.has('win32') == 1
+
+--- Checks whether a given path exists and is a directory.
+---@param filename string path to check
+---@return boolean
+local function is_dir(filename)
+  local stat = uv.fs_stat(filename)
+  return stat and stat.type == 'directory' or false
```
Comment (justinmk) [inline runtime/lua/vim/lsp/_transport.lua:11]: Looks like a vote in favor of `vim.fs.isdir` :)

**Comment by justinmk** [comment]
This makes sense and maps to the layout in `test/client/` (though the file names there could be improved, which I may do as part of https://github.com/neovim/neovim/pull/30384 )

File: runtime/lua/vim/lsp/_transport.lua [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+--- @return vim.lsp.rpc.Transport.Spawn
+function Spawn.new()
+  return setmetatable({}, { __index = Spawn })
+end
+
+--- @param cmd string[] Command to start the LSP server.
+--- @param extra_spawn_params? vim.lsp.rpc.ExtraSpawnParams
+--- @param on_read fun(err: any, data: string)
+--- @param on_exit fun(code: integer, signal: integer)
+function Spawn:spawn(cmd, extra_spawn_params, on_read, on_exit)
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply runtime/lua/vim/lsp/_transport.lua:33]: I just wanted to have something more distinct when compared to `connect()`. One class is for spawning an internal process, and the other is connecting to an address, so I felt `run()` was a little too generic. But I guess `run` does work if you consider Python's `subprocess.run()`.

File: runtime/lua/vim/lsp/_transport.lua (hunk 1 above)
Comment (lewis6991) [reply runtime/lua/vim/lsp/_transport.lua:11]: It does overlap with `vim.fn.isdirectory`, but that function returns `0|1`, so I'd be in favour of `vim.fs.isdir`.

**Review by justinmk — APPROVED** [review]
LGTM, but probably needs approval from someone who worked closely on this code

**Review by mfussenegger — APPROVED** [review]

**Pull request merged**